Add a "new record" button at the bottom of grid and detail views

manuhabitela opened the pull request:
## Context

See https://github.com/gristlabs/grist-core/issues/366.

Lucie's mockups: https://www.figma.com/design/wcpetFt6aOKzTszcvPPWLQ/%5B05%2F24%5D-Grist-Design?node-id=3744-62552&t=MhknMbyFwd1RAc9n-0

## Proposed solution

Very early wip that lead me to some questioning:

### Conditions to show the button

The show/hide condition of the button certainly needs refinement. I mostly rely on the view's `enableAddRow` boolean, and show the button only when the view is focused, but that's pretty much it.

### Intended behavior on click

I'm not sure about the intended behavior.  

I made it so that clicking the button actually adds a new record, and not just focuses the "new record" blue row. I felt it matched more what we would expect, at least in the grid view. Otherwise, clicking the "new record" multiple times would not actually add records but just stay focused on the last line.

But I feel this current behavior is a bit misleading when using the detail view. Maybe in that case we would want to just show the card with blueish inputs?

### Button wording

I added the possibility to have different wording on the button depending on the view. For now it renders "New record" on the grid view and "New card" on the detail view.

I did that because it feels different wording for the detail view would maybe better, at least in some other languages than English. In French for example, translating "record" is a bit long and generic.

What do you think?

## Related issues

Fixes https://github.com/gristlabs/grist-core/issues/366.

## Has this been tested?

<!-- Put an `x` in the box that applies: -->

- [ ] 👍 yes, I added tests to the test suite
- [x] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this is not relevant here
- [ ] 🙋 no, because I need help <!-- Detail how we can help you -->

## Screenshots / Screencasts

Here is a preview: 

_Note: the preview is not up-to-date, as I updated the code to show we can have different wording for different views. Right now the button renders "new card" when focusing the detail view instead of "new record"_

https://github.com/user-attachments/assets/efb33849-b90e-458c-acf6-0a930321f9f4

manuhabitela commented on app/client/components/buildViewSectionDom.ts:
```diff
@@ -129,7 +130,12 @@ export function buildViewSectionDom(options: {
         ),
         dom.cls((use) => 'viewsection_type_' + use(vs.parentKey)),
-        viewInstance.viewPane
+        viewInstance.viewPane,
+        dom.domComputed(use =>
+          (use(viewInstance.viewSection.hasFocus) && use(viewInstance.enableAddRow)),
+          (showNewRecordButton) =>
+            showNewRecordButton ? newRecordButton(viewInstance) : null
+        )
```
This is out of the `viewPane` mostly because some views have `overflow: hidden` on their wrapper and since we want the button to be a bit out of the view it needs to be outside the wrapper.

vviers commented:
My two cents : on a new table that's empty, the new record can be a bit tricky to find (it took me a minute 🙈 )

![image](https://github.com/user-attachments/assets/fc813f19-4ae5-4426-8560-4c12b13adab9)

vviers commented:
Also, I'm unclear as to what the different between 1/ the new "Add card" button and  2/ the old "+" on top of the card view

It seems to me that they do slightly different things (1/ creates a new empty record whereas 2/ doesn't quite yet create the record but gives you the option to do it), but I don't know that they should ? 🤔 

![image](https://github.com/user-attachments/assets/992dc276-57ed-42bd-adc3-3c4bcca7f011)

manuhabitela commented:
> Also, I'm unclear as to what the different between 1/ the new "Add card" button and 2/ the old "+" on top of the card view is
> 
> It seems to me that they do slightly different things (1/ creates a new empty record whereas 2/ doesn't quite yet create the record but gives you the option to do it), but I don't know that they should ? 🤔

Yes the behavior is, for now, purposefully different (see my questioning part).

I feel for the card view it should stick to always doing the original behavior (number 2). But I feel on the grid view current behavior is better (number 1)?

lusebille commented:
> My two cents : on a new table that's empty, the new record can be a bit tricky to find (it took me a minute 🙈 )
> 
> ![image](https://private-user-images.githubusercontent.com/30295971/410969678-fc813f19-4ae5-4426-8560-4c12b13adab9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA0MDQyNjAsIm5iZiI6MTc0MDQwMzk2MCwicGF0aCI6Ii8zMDI5NTk3MS80MTA5Njk2NzgtZmM4MTNmMTktNGFlNS00NDI2LTg1NjAtNGMxMmIxM2FkYWI5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI0VDEzMzI0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcwZWFjYWRmZTY4OWM0MzM3MzczZjBiMGVlZGI0YjJiZTg4MDlmZGNiNTk5MWI0OGYwYTIzZTYwNDMzMDYxNDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-0TUHw-aFVTuxYrzdGxSjGc3PrRxTtS7SP5ZIHJiAho)

I think the 'add new record' button is useful if you've got lot of lines (so in lot of cases), we keep the actual behavior too so when you're on a new table, i you're filling a first line , a new one is automatically created, so in that case you don't need to use the'new record button' but it's not a problem because they're two ways to do it

lusebille commented:
> Also, I'm unclear as to what the different between 1/ the new "Add card" button and 2/ the old "+" on top of the card view is
> 
> It seems to me that they do slightly different things (1/ creates a new empty record whereas 2/ doesn't quite yet create the record but gives you the option to do it), but I don't know that they should ? 🤔
> 
I'm not fully sure about the actual behavior with the actual '+' (for exemple a 'add record' appear and user doesnt know if it's a button or a title, and the '+' is not very visible)  , maybe adding 'add new record' button may sound not very useful for the card view but if user use 'add new record' for table view to create 5 new empty lines for example , he will find 5 new empty cards too if he's also using a card view, so better to let him the possibility to create it on the card view with a button

dsagal commented:
I added a couple of our templates to the preview, to make it easier to play around with slightly more realistic data:
- https://grist-manuhabitela-grist-core-issue-366-new-record-button.fly.dev/i8oDc8KyGvsH/Lightweight-CRM
- https://grist-manuhabitela-grist-core-issue-366-new-record-button.fly.dev/aPZas6h9DydN/Class-Enrollment/p/7

One early thought: when you select a contact in the CRM example, the common next step is to add a new interaction, but there is no indication how to do it -- no "New record" button for interactions. One still needs to know to click into that widget (which means clicking on an existing record, when there are many rows) just to see the option to add a new record.

Another thought is that in a list+card arrangement as in the CRM example, adding a new contact in the table would normally be followed by entering the contact data in the linked card. In a custom app, the natural flow would be a page-wide "Add Contact" button, which focuses the Card widget every time. This is a vague thought, but perhaps someone has ideas.

I think the unifying idea of these observations is that when there are linked widgets, the behavior of "New record" should ideally depend on the shape of linking.

vviers commented:
> I think the unifying idea of these observations is that when there are linked widgets, the behavior of "New record" should ideally depend on the shape of linking.

This makes sense to me. Anecdotally, I've seen people use the ["Action Button"](https://github.com/gristlabs/grist-widget/tree/master/actionbutton) custom widget for this exact purpose (create a new record in a way that makes sense given the page that you are in, e.g. "add a new contact").

emanuelegissi commented:
@lusebille 
Here is what my `New record` action button looks like (quite suboptimal).

See this worflow management example I am working on:
https://docs.getgrist.com/iFLERrF5h1rd/Approval-workflow?utm_id=share-doc

Beware: it is still in development!
I still have a problem with record duplication... no easy way with the current plugin api.

dsagal commented:
@manuhabitela , for the current implementation, I think it addresses only a subset of the confusion people have, and in some cases creates new confusions. So we wouldn't be comfortable releasing it to production as a default behavior.

But doing user-testing and getting real user feedback on this would be great, so we are OK releasing it to production behind a URL parameter (as we discussed elsewhere, something simple like check `new URLSearchParams(location.search)` directly). This would allow you to test with users on real-life documents and get more feedback.

We can then decide whether to keep this approach, if we need to give the user an option to turn it on or off, or if we want to do something different.

manuhabitela commented:
Oh alright!

What I had in mind after we talked about it a few weeks ago, and I think already started coding:

- we can share a url like `https://docs.getgrist.com/?experiment=new-record-button`
- going on this url shows a modal like:

> **"New record" button**
>
> You are about to enable this experimental feature: _"New record" button_.
>
> To stop using the feature, you can return on this URL at any time: https://docs.getgrist.com/?experiment=new-record-button
> 
> [Cancel] / [Enable the feature]

- clicking _Enable the feature_ would redirect to homepage and have the feature flag enabled in the app[1]
- going back on the URL would show the same modal but explaining that you are about to disable the feature.

[1] What I was not sure, and don't remember, was if the user-feature stuff was already something completely coded, allowing to easily add new features tied to users, saved in their preferences. If not, I imagined I could just rely on localStorage and make it explicit in the modal that the feature would be enabled only on current machine.

What do you think?

manuhabitela commented:
I have to say I'm not _that_ convinced with only keeping a variable in the state, since full page loads happen between documents list and document page, it quickly narrows the possibilities to what you just said: a specific testing session on one given document.

I feel the dialog + keeping state in localstorage is not _that_ much more code to maintain, stays rather simple (only client-side), and a good compromise between "only keep in js state for current doc" vs "keep it in user prefs server-side".

It easily allows to globally share a link saying "hey, you can enable this experimental feature with this url and then work as usual on your docs to get a feel for it".

So if you don't object I'd rather implement that. I actually checked earlier and I was almost done implementing the dialog + localstorage save, it's not that much code and it's pretty simple :innocent:

manuhabitela commented:
@dsagal what do you think of this implementation? _Note: tests tell me I missed a little something but I have to go right now_ :clown_face:

Here is a preview of what it does (@lusebille):


https://github.com/user-attachments/assets/913a6c52-3571-4aff-ae6f-d7d68386762f

lusebille commented:
> https://grist-manuhabitela-grist-core-issue-366-new-record-button.fly.dev

Hello, It's a good thing to display a message to user with a way to cancel the feature but I wonder if this link shouldn't be always findable somewhere ( in a tooltip next to the button ?) Because if user miss to copy it he can't cancel the feature ?

manuhabitela commented:
Not a fan of having some UI next to the actual thing we want to test. I'd be more inclined in having the enabled experiments shown in the user settings?

But since this is kind of a hidden feature/not really meant to be used for loads of stuff (for now at least), I'd say let's not take more dev time on this and ship as is? This is meant to be used in a rather controlled way with people, sharing the feature toggle URL with specific people, I feel it's rather ok to not expose this too much in the UI.

I can change the wording to tell to better tell the URL is shown only here?

lusebille commented:
> Not a fan of having some UI next to the actual thing we want to test. I'd be more inclined in having the enabled experiments shown in the user settings?
> 
> But since this is kind of a hidden feature/not really meant to be used for loads of stuff (for now at least), I'd say let's not take more dev time on this and ship as is? This is meant to be used in a rather controlled way with people, sharing the feature toggle URL with specific people, I feel it's rather ok to not expose this too much in the UI.
> 
> I can change the wording to better tell the URL is shown only here?

Yes it's obviously better in settings but I knew it's more dev so for this reason I was looking for a quick way to do it , I was thinking into a information message below the header too , my point is most part of users doesn't read pop up so I may think they won't even realize they have to copy something to cancel action, but we might put this link too in the message for recruiting them

fflorent commented on static/locales/fr.client.json:
```diff
@@ -1946,5 +1946,23 @@
+        "New record": "Nouvelle ligne"
+    },
+    "Experiments": {
+        "{{- experiment}} disabled.": "{{- experiment}} : désactivé.",
```
Same as for the other PR, I fear that Weblate will be lost if we start translating strings from the repo.

manuhabitela replied:
I'll just remove those and go through weblate later :+1:

fflorent reviewed (commented):
Tested, this looks good.

It is experimental, so a feedback I have is not blocking any at all: I would rather expect the "New record" button to add a new line below the selected line. But that's arguable and it should be discussed with users using this experiment :)

dsagal commented on app/client/ui/Experiments.ts:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    // If the user just confirmed the feature toggle, feedback the change and force a page reload
+    hasConfirmedModal.addListener((val) => {
```
I generally recommend avoiding `addListener`, and in this case I don't think you even need an observable. You could have a method `_showConfirmation()` that shows the first dialog, and `_showReloadPrompt()` for the second dialog, and instead of calling `hasConfirmedModal.set(true)`, you could just call `this._showReloadPrompt()`. (I may be missing some subtlety, but it looks like that's all that's happening.)

manuhabitela replied:
you are totally right, don't know what I was thinking here :)

dsagal commented on app/client/ui/Experiments.ts:
```diff
@@ -0,0 +1,189 @@
+      if (!val) {
+        return;
+      }
+      modal(
```
I am pretty sure you could use `confirmModal` here too (with `hideCancel: true`, since there's just one button). There isn't a big difference though, so up to you.

manuhabitela replied:
Oh sure, why not, thanks.

dsagal commented on app/client/ui/NewRecordButton.ts:
```diff
@@ -0,0 +1,54 @@
+    dom('span', t(translationString)),
+    dom.on('click', () => {
+      if (view.onNewRecordRequest) {
+        view.onNewRecordRequest();
```
We've started using the neat new JS/Typescript shorthand: `view.onNewRecordRequest?.()`

dsagal commented on app/client/ui/NewRecordButton.ts:
```diff
@@ -0,0 +1,54 @@
+  const translationString = translationStrings[viewType as keyof typeof translationStrings]
+    || 'New record';
+  return cssNewRecordButton(
+    iconSpan('Plus'),
```
just `icon('Plus')` should be enough ("span" shouldn't make a difference because it's in a flexbox).

dsagal commented on app/client/ui/NewRecordButton.ts:
```diff
@@ -0,0 +1,54 @@
+    || 'New record';
+  return cssNewRecordButton(
+    iconSpan('Plus'),
+    dom('span', t(translationString)),
```
Not sure this works as expected with tools that extract strings for translation -- maybe better to apply the `t()` wrapper where the literals are defined (in lines 11, 12, 27)?

dsagal commented on app/client/ui/NewRecordButton.ts:
```diff
@@ -0,0 +1,54 @@
+  align-items: center;
+  gap: 6px;
+
+  /* 16px on the plus icon is blurry, 17px is sharp, needs more test. */
```
This is surprising -- the "+" icon in the "Add new" button always visible in the upper left of Grist has 16x16 size.

manuhabitela replied:
Oh you are right. But now that you mention it, I realize the "add new" button icon is also blurry for me with default zoom on various OS/browsers actually. I'll just remove these couple css rules to match current behavior.

dsagal reviewed (commented):
Thanks @manuhabitela , sorry it took a while to get to the review! (And I know there is a bigger one waiting for me)

manuhabitela reviewed (commented):
Thank you both for the review, currently making changes, will let you know.

fflorent commented:
@manuhabitela Could you rebase your work?

@georgegevoian Could you please notify us when you request a rebase, otherwise we may not be aware of that :). Thanks in advance!

manuhabitela committed:
pull request feedbacks. Thanks Dmitry and Florent

manuhabitela committed:
(ui) "new record button" experimental feature

- add logic to enable the toggling of "experiments" via URL search
params
- first experiment of this sort is a "new record button" that is visible
at the bottom left of a grid or detail view.

manuhabitela commented:
I have read the CLA Document and I hereby sign the CLA

georgegevoian reviewed (approved):
Thanks @manuhabitela!

vviers commented:
Apparently the button appears on custom widgets, which may not be the desired behavior !

DavidFabijan commented:
It does appear on custom widgets, and the usefulness there really depends a lot on the situation. For things like tables and cards I can see it being very useful. For stuff like calenders also ... though it would have to be handled in a nicer way (like I mentioned in an earlier comment). For other things like HTML viewers, charts, etc. it makes no real sense at all. Perhaps an interesting question would be what to do with the notepad/markdown apps, those might benefit from the button. I do have a use case where and Add new button for a notepad widget would be handy. But it is quite a niche use and probably not good practice, so we might actually do better without a button for those widgets.

manuhabitela commented:
Very sorry that I missed the feedback here. I opened a new PR to remove the button when printing, and only show the button in tables and cards widgets.
